extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/JdbcBasePersistenceImpl.java

```diff
@@ -0,0 +1,619 @@
+        ? Collections.emptyList()
+        : results.stream().map(ModelEntity::toEntity).collect(Collectors.toList());
+  }
+
+  @Nonnull
+  @Override
+  public List<PolarisChangeTrackingVersions> lookupEntityVersions(
+      @Nonnull PolarisCallContext callCtx, List<PolarisEntityId> entityIds) {
+    Map<PolarisEntityId, ModelEntity> idToEntityMap =
+        lookupEntities(callCtx, entityIds).stream()
+            .collect(
+                Collectors.toMap(
+                    entry -> new PolarisEntityId(entry.getCatalogId(), entry.getId()),
+                    ModelEntity::fromEntity));
+    return entityIds.stream()
+        .map(
+            entityId -> {
+              ModelEntity entity = idToEntityMap.getOrDefault(entityId, null);
+              return entity == null
+                  ? null
+                  : new PolarisChangeTrackingVersions(
+                      entity.getEntityVersion(), entity.getGrantRecordsVersion());
+            })
+        .collect(Collectors.toList());
+  }
+
+  @Nonnull
+  @Override
+  public List<EntityNameLookupRecord> listEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      @Nonnull PolarisEntityType entityType) {
+    return listEntities(
+        callCtx,
+        catalogId,
+        parentId,
+        entityType,
+        Integer.MAX_VALUE,
+        e -> true,
+        EntityNameLookupRecord::new);
+  }
+
+  @Nonnull
+  @Override
+  public List<EntityNameLookupRecord> listEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      @Nonnull PolarisEntityType entityType,
+      @Nonnull Predicate<PolarisBaseEntity> entityFilter) {
+    return listEntities(
+        callCtx,
+        catalogId,
+        parentId,
+        entityType,
+        Integer.MAX_VALUE,
+        entityFilter,
+        EntityNameLookupRecord::new);
+  }
+
+  @Nonnull
+  @Override
+  public <T> List<T> listEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      PolarisEntityType entityType,
+      int limit,
+      @Nonnull Predicate<PolarisBaseEntity> entityFilter,
+      @Nonnull Function<PolarisBaseEntity, T> transformer) {
+    Map<String, Object> params = new HashMap<>();
+    params.put("catalog_id", catalogId);
+    params.put("parent_id", parentId);
+    params.put("type_code", entityType.getCode());
+
+    // Limit can't be pushed down, due to client side filtering
+    // absence of transaction.
+    String query =
+        JdbcCrudQueryGenerator.generateSelectQuery(
+            ModelEntity.class, params, null, null, "last_update_timestamp");
+    List<ModelEntity> results = datasourceOperations.executeSelect(query, ModelEntity.class);
+    return results == null
+        ? Collections.emptyList()
+        : results.stream()
+            .map(ModelEntity::toEntity)
+            .filter(entityFilter)
+            .map(transformer)
+            .limit(limit)
```

Comment:
Would it be more efficient to push `limit` down to the database?

Reply:
I'm not sure how it can be a problem for atomicity... In any case, I see now that there's a java `Predicate` in this method. That can certainly not be pushed down easily... but we should apply the `Predicate` in chunks and avoid loading the whole Result Set into a `List`

Reply:
sure thing added `result.Next()` it converts as per the transformation and then apply Predicates and does till it collects till asked Limit is reached, this way memory would be contained, though would be slow.